apiserver/facades/client/application/charmstore.go

```diff
@@ -248,11 +249,32 @@ func openCSClient(csAPIURL string, args params.AddCharmWithAuthorization) (*cscl
 	return csClient, nil
 }
 
-func checkMinVersion(ch charm.Charm) error {
+func checkMinVersion(ch charm.Charm, caasVersion *version.Number) (err error) {
+	// check Juju min version.
 	minver := ch.Meta().MinJujuVersion
 	if minver != version.Zero && minver.Compare(jujuversion.Current) > 0 {
 		return minVersionError(minver, jujuversion.Current)
 	}
+
+	// check caas min version.
+	charmDeployment := ch.Meta().Deployment
+	if caasVersion == nil || charmDeployment == nil || charmDeployment.MinVersion == "" {
+		return nil
+	}
+	if len(strings.Split(charmDeployment.MinVersion, ".")) == 2 {
+		// append build number if it's not specified.
+		charmDeployment.MinVersion += ".0"
+	}
+	minver, err = version.Parse(charmDeployment.MinVersion)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if minver != version.Zero && minver.Compare(*caasVersion) > 0 {
+		return errors.NewNotValid(nil, fmt.Sprintf(
+			"charm's deployment min version(%q) is higher than remote cloud version (%q)",
```

Comment:
This method is only intended to check the juju version and adding the k8s version check makes it bloated and a bit messy. 
We really want a separate checkMinK8sVersion() method.

The k8s error message should be something like:
"charm requires a minimum k8s version of %v but the cluster only runs version %v"